### ISSUE TITLE
Problem: Adding a 2nd (or later) type parameter to a class breaks clients

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#283]
and incrementally migrating is not easy; it must be done atomically with the type-parameter-increasing change.

This is not a problem when adding the _first_ type paramter:

```dart
class C {}

// This is a non-breaking change:
class C<T> {}

// This is also non-breaking:
class C<T, U, V> {}
```

However, when adding any type parameters beyond the first, the change is breaking for any clients which specify any type arguments:

```dart
class C<T> {}

// This is a breaking change:
class C<T, U> {}

// For code such as the following:
new C<int>();
C<int> c;
class D extends C<int> {}
```

Introducing a second class while renaming the original class is one way to "deprecate" the previous number of type parameters, and expose the "new" number of type parameters:

```dart
class C<T> {
  T f1;
  dynamic f2;
}

// Non-breaking change:
class C_2<T, U> {
  T f1;
  U f2;
}

@Deprecated("Now with more type parameters! Use C_2!")
class C<T> extends C_2<T, dynamic> {}
```

You can ship this non-breaking change, announcing the deprecation, allowing clients to migrate, and later ship a breaking change where you remove `C`.

The big problem with this is that no one wants to rename classes. If you look at functions, it is a breaking change to introduce or remove required parameters (or remove or rename named parameters), but this is mitigated by making optional parameters available. Or perhaps the purpose of a function has changed enough with the changed parameters that you can offer it as a new function with a new name, and keep the old function around, marking it `@Deprecated()`.

Renaming a class seems like a much bigger and unfortunate change. Classes often do not involve more than two words; typically a few nouns, rather than an English predicate expression. "ContainerViewModel" means the Model of a View of a Container, which should not be changed based on the number of type parameters.

**Comment by eernstg** [comment]
Would the following approach be sufficiently helpful? (It relies on generalized type aliases, but that feature is coming, it is just sitting in the pipeline a bit longer than we thought because of other features with a higher priority).

Before the change we have this (I'll focus on the change from one to two type parameters because that one embodies the harder issue):

```dart
// Library 'useful_stuff.dart'.
class C<X> { ... }

// Library 'client.dart'.
import 'useful_stuff.dart';
... // Lots of occurrences of `C` in here.
```

Then we change `C` to have two type arguments, and we'd like to have a per-library migration. We rename the library 'useful_stuff.dart' to 'new_useful_stuff.dart' and provide a new 'useful_stuff.dart' library to make the transition easier:

```dart
// Library 'new_useful_stuff.dart'.
class C<X> { ... }

// Library 'useful_stuff.dart'.
import 'new_useful_stuff.dart' as p show C;
export 'useful_stuff.dart' hide C;
typedef C<X> = p.C<X, dynamic>;
```

Client libraries stay unchanged, but their imports will now refer to the new "bridge" library 'useful_stuff.dart' rather than the library that actually declares `C`.

Each client can then make the choice to change its import to `new_useful_stuff.dart' and adjust the code to provide the new type argument in useful ways.

Of course, if the name 'useful_stuff.dart' must remain unchanged then we could also use a new name for the bridge library, and then clients would have to adjust their imports. That trade-off could be OK, because it is nearly a textual operation (a global search and replace).

**Comment by srawlins** [comment]
This definitely offers an alternative to class renaming.

If the class is defined in a `/src/` library which is already only accessed via a public `export`, then this solution may be more awkward than the class renaming.

```dart
import 'package:foo/foo.dart';
```

becomes

```dart
import 'package:foo/foo.dart'; // old C is hidden by foo.dart
import 'package:foo/foo_transitional_c.dart';
```

**Comment by eernstg** [comment]
Of course, this might not work in practice (if there are several classes with changes like this, and transitioning might not happen all at once, etc), but in the simple case where we only change `C` it seems like `foo_transitional_c.dart` might as well export the same name space as `foo.dart` used to export (that is, the `typedef`'ed `C`, plus all other things with no changes), and clients would then only need to change

```dart
import 'package:foo/foo.dart';
```

to

```dart
import 'package:foo/foo_transitional_c.dart';
```

When they are ready to update that particular library to use the new `C` they'll switch back to `foo.dart` and fix the code.

Again, the naming could be swapped such that clients don't have to change _anything_ initially, if it has top priority to avoid changes in client code, and if it is acceptable to use a new name for the library when the transition is over.

**Comment by munificent** [comment]
> This is not a problem when adding the first type parameter:

Technically, it can be a problem here too if the first type parameter has a bound that inference can't satisfy, because then using the raw type becomes an error. More generally, for users that enable "strict raw types", adding a type parameter will become an error.

C# handles this like it handles adding parameters to methods: it lets you overload. You can define multiple classes with the same name but different type parameter arity:

```c#
class C {}
class C<T> {}
class C<T1, T2> {}
```

**Comment by caseycrogers** [comment]
Your analysis/proposal is a lot more expansive than mine-my assumption/use case is that the default value would be used only if the default type had also been used. eg `Foo<num>()` would and should be a compile time error because `default_value` is not a num. `Foo()`, or `Foo<num>(bar: 4)` would also be valid. When a user specifies `bar`, `T` is inferred by the type of `bar`. To me it'd be intuitive for `T` to be inferred from the type of `'default_value'` when `bar` is omitted (the first of the two proposals in my comment above).

That said, your proposal (assuming default values were first allowed to be non const) sounds like it could be useful. Eg if you're creating a class for storing and manipulating data types, you might have a list of types with supported zero-values and want to allow users to skip specifying an explicit value for any class with a valid zero value. I don't think there's a way to get around the unsafe `computeDefaultValue` code though....

**Comment by eernstg** [comment]
It's an interesting idea that the default value is essentially considered to exist only when the typing situation allows it to be used (in all other situations the parameter would be considered to be `required`).

It's not trivial to define that approach, however, because many types are only known by an upper bound:

```dart
class A<X> {
  void m([X x ?= 42]);
}

void main() {
  var a = A<int>();
  a.m(10); // OK.
  a.m(); // OK, `42 is X`.
  A<num> a2 = A<double>();
  a2.m(); // Run-time error, `42 is X` does _not_ hold!
}
```

So we can basically only rely on having a default value when we have a _lower_ bound `L` for the value of the type variable `X`, and that is just about the same thing as knowing the exact type. It would also work if we introduce declaration site variance and `X` is contravariant:

```dart
// Requires `--enable-experiment=variance`.

class A<in X> {
  void m([X x ?= 42]);
}

void main() {
  var a = A<int>();
  a.m(10); // OK.
  a.m(); // OK, `42 is X`.
  A<double> a = A<num>(); // Note that A<num> <: A<double>.
  a.m(); // Compile-time error, `42 is X` is not known to hold.
}
```

So it might actually work quite well together with sound variance. Interesting! ;-)

**Comment by mateusfccp** [comment]
@eernstg I'm considering this as a viable alternative to introduce phantom types to a type without breaking it.

In this case, it wouldn't require any default values, but its usefulness would also be reduced.

Aside from the default values problem, is there any other problem that you see that may arise from this issue?

**Comment by eernstg** [comment]
I don't think this issue presents a specific proposal, it's a 'request' rather than a 'feature' issue. Of course, some comments may contain the beginnings of a proposal, including type parameters with default values, or workarounds, like the use of type aliases, and each of those would have pros and cons.

By the way, I don't think anybody mentioned that the additional type parameter could be added in a subtype:

```dart
// Current version.
class A<X> {
  final X x;
  A(this.x);
}
```

```dart
// Next version has a phantom type.

class A<X> {
  factory A(X x) = A2<X, String>; // Assuming a "default" of `String` for `Y`.
}

class A2<X, Y> implements A<X> {
  final X x;
  A2(this.x);
}
```

This would work as long as the constructor change is acceptable: In the first version, `A` has a generative constructor, which means that it is possible to create subclasses of it. In the second version we use a factory constructor, which would break the subclasses, but on the other hand we do get an instance of `A2` which should work as a replacement for `A`, and it will use the specified default value for `Y`. This may or may not be a workaround that actually works, but it should be pretty safe if the class is `interface` (because subclasses declared in the same library can be fixed, and subclasses in other libraries cannot exist).

**Comment by mateusfccp** [comment]
> I don't think this issue presents a specific proposal, it's a 'request' rather than a 'feature' issue.

Yeah, I am aware.

I asked because I thought of writing a proposal with what I had in mind, but I didn't consider the default value question.

For my specific use case, it's irrelevant. As I was thinking about _phantom types_, no value is used. In this case, we could simply ignore this and decide that:

- A type parameter with a default value must be an interface (no implementation);
- A type parameter with a default value does not need to be an interface, but can't use the type parameter with default value as a `final` field.

I don't know if this would be enough, I didn't think it through.

However, as I said, I don't know if this kind of proposal would bring much value, so I'm trying to identify which potential problems arise from this issue.

**Comment by mateusfccp** [comment]
@eernstg wrote:

> Does that work for your case?

Hmm, not really.

This would work if the client wouldn't ever want a different phantom type, but let's introduce a new class in the client.

```dart
// --- Library.

class A { /* ... */ }

// --- A client.

class B extends A {}

class C extends A {}
```

Now, I want to introduce the phantom type to `A`. I want `B` to use the "default" value, while `C` should use a different type.

```dart
// --- Library.

class NewA<Tag> {}

typedef A = NewA<DefaultTag>;

abstract interface class Tag {}

final class DefaultTag implements Tag {}

final class OtherTag implements Tag {}
```

Now, even if `B` does not break, for `C` I will have to implement `NewA` instead of `A`.

```dart
// --- A client.

class B extends A {} // Does not break, uses `DefaultTag` correctly

class C extends NewA<OtherTag> {} // Can't extend `A`, have to extends `NewA`
```

**Comment by eernstg** [comment]
Right, we would be able to make the transition more smooth with default values for formal type parameters.

On the other hand, it seems likely that you're changing the semantics of the class `C`, because (presumably) the meaning of `A` should be the same as the meaning of `NewA<DefaultTag>`, and you want to use `extends NewA<OtherTag>` with `C`. So it doesn't demonstrate a need to use a different superclass than `A` with a class that should work exactly as it has worked until now.

Anyway, I have no doubts that default values of formal type parameters can make some evolutionary changes to an API easier to roll out.
